poly/src/polynomial/mod.rs

```diff
@@ -1,164 +1,93 @@
-//! Work with sparse and dense polynomials.
-
-use crate::{EvaluationDomain, Evaluations};
-use ark_ff::{FftField, Field};
-use ark_std::{borrow::Cow, convert::TryInto, vec::Vec};
-use DenseOrSparsePolynomial::*;
-
-mod dense;
-mod sparse;
-
-pub use dense::DensePolynomial;
-pub use sparse::SparsePolynomial;
-
-/// Represents either a sparse polynomial or a dense one.
-#[derive(Clone)]
-pub enum DenseOrSparsePolynomial<'a, F: Field> {
-    /// Represents the case where `self` is a sparse polynomial
-    SPolynomial(Cow<'a, SparsePolynomial<F>>),
-    /// Represents the case where `self` is a dense polynomial
-    DPolynomial(Cow<'a, DensePolynomial<F>>),
-}
+//! Modules for working with univariate or multivariate polynomials.
+
+use ark_ff::Field;
+use ark_std::{
+    fmt::Debug,
+    hash::Hash,
+    ops::{AddAssign, Index, SubAssign},
+    vec::Vec,
+};
+use rand::Rng;
+
+pub mod multivariate;
+pub mod univariate;
+
+/// Describes the common interface for univariate and multivariate polynomials
+pub trait Polynomial<F: Field>:
+    Sized
+    + Clone
+    + Debug
+    + Default
+    + Hash
+    + PartialEq
+    + Eq
+    + Send
+    + Sync
+    + for<'a> AddAssign<&'a Self>
+    + for<'a> AddAssign<(F, &'a Self)>
+    + for<'a> SubAssign<&'a Self>
+{
+    /// The domain of the polynomial.
+    type Domain: Sized + Clone + Ord + Debug + Sync;
```

Comment:
Let's come up with a better name for the associated type; this might cause confusion with the `EvaluationDomain` related-things.